### PR TITLE
Non-invasive introduction of replaceAt and replace methods in Object3D (src and docs)

### PR DESCRIPTION
**Pull request by EliasHasle**

### CODE REVIEW

**Comment by EliasHasle** [comment]
[Original request for new feature](https://github.com/mrdoob/three.js/issues/8325)

**Comment by EliasHasle** [comment]
Original feature request: #8325
@MasterJames Related pull request: #8406

**Comment by WestLangley** [comment]
Did you test these cases?

What if `newChild` equals `oldChild`?
What if `newChild` is already a child, at a different index?
What if `newChild` is a child of `oldChild`?
What if `index` is out-of-bounds?

Perhaps there are additional cases to test.

**Comment by MasterJames** [comment]
Please look again at my revised version. I'm interested to address/explain any concerns and or learn from the discussion thank you.

**Comment by EliasHasle** [comment]
@WestLangley Thanks for the feedback. You are right. It is breakable in its current form. All those cases should be considered, and maybe tested for. Old parent = this will fail badly, but is easy to check for. I don't have time right now, sorry. You are welcome to try fixing it, orelse I will do it another day. :-)

**Comment by EliasHasle** [comment]
@MasterJames Your code is a lot better now. I was worried it was too broad/ambitious. Therefore my PR. I wish you luck. No problem if your patch "wins", as long as it's the best one. :-)

**Comment by MasterJames** [comment]
Yes thank you for the vote of confidence and continue support & understanding.

**Comment by EliasHasle** [comment]
@WestLangley 
> What if newChild equals oldChild?
> What if newChild is already a child, at a different index?
> What if index is out-of-bounds?
Checks added.

> What if newChild is a child of oldChild?
No problem. newChild is removed from oldChild, and added to _this_ at oldChild's index. oldChild is not destroyed by this method. :-)

I feel obliged to admit that this code is entirely untested. I have little experience with testing, actually, so I wouldn't trust my tests anyway...

**Comment by EliasHasle** [comment]
Here is a little script I have just used for testing normal operation:
```javascript
			//Save (preload) a bunch of different hat objects in hats.
			hats = {};
			hatColors = ["Red", "Green", "Blue", "White", "Cyan", "Magenta", "Yellow", "Black"];
			var hat, color;
			for (var i=0; i<hatColors.length; i++) {
				hat = new THREE.Object3D();
				color = hatColors[i];
				hat.name = color + " hat";
				hats[color] = hat;
			}
			
			//
			shirts = {};
			shirtBrands = ["Armani", "Gant", "Sand"];
			var shirt, brand;
			for (var i=0; i<shirtBrands.length; i++) {
				shirt = new THREE.Object3D();
				brand = shirtBrands[i];
				shirt.name = brand + " shirt";
				shirts[brand] = shirt;
			}

			//
			pants = {};
			pantShapes = ["No", "Straight", "Skinny", "Boot-cut"];
			var pant, shape;
			for (var i=0; i<pantShapes.length; i++) {
				pant = new THREE.Object3D();
				shape = pantShapes[i];
				pant.name = shape + " pants";
				pants[shape] = pant;
			}
			
			//and so on...
			
			//Now I initialize an outfit:
			var outfit = new THREE.Object3D();
			outfit.add(hats["Red"]); //index 0 is for the hat
			outfit.add(shirts["Armani"]); //index 1 is for the shirt
			outfit.add(pants["No"]); //index 2 is for the pants
			
			//Then the customer wants to try a new combination:
			outfit.replaceAt(0, hats["Green"]);
			outfit.replaceAt(1, shirts["Gant"]);
			outfit.replaceAt(2, pants["Straight"]);
			//It is no problem here to repeat a line:
			outfit.replaceAt(2, pants["Straight"]);
			
			//He can also choose to do like this, _if_ he knows what clothes are on, 
			//and doesn't want to bother with what part of his body has each piece on:
			//(but it will technically be slower than replaceAt)
			outfit.replace(hats["Green"], hats["Blue"]);
			outfit.replace(shirts["Gant"], shirts["Sand"]);
			outfit.replace(pants["Straight"], pants["Skinny"]);

			//This should now print an array of objects named:
			//"Blue hat", "Sand shirt", "Skinny pants"
			//(and it does.)
			console.log(outfit.children);
```

**Comment by EliasHasle** [comment]
@mrdoob @WestLangley OK, I believe this is close to optimal now. I request feedback. :-)

Should I add some more features, like @MasterJames has done in #8406 ? I don't mean all, but would you like to have simple, conservative implementations of addAt and removeAt in the same PR?

**Comment by EliasHasle** [comment]
In case someone didn't notice, I updated the first post of #8325.
TLDR^2:

> This PR is amazing and super-useful, so please merge it asap.

TLDR:

> I have found this low-level logic useful and natural, and think others can benefit from it too. In some cases, maybe extreme ones, it matters for performance. It is only a small enhancement, but.at least nothing gets broken...

**Comment by WestLangley** [comment]
I am not sure why you allow negative indices, and insert the object at an index different from the one the user requested.

Sorry. As I indicated a month ago, I do not believe `replaceAtIndex( index, newChild )` is a method we should be implementing. It is too risky.

Furthermore, and most importantly, indices change as objects are inserted and removed.

Such a method is appropriate as an application-level method where the use-case is restricted and under developer control.

**Comment by EliasHasle** [comment]
Thanks for your answer. :-)

> I am not sure why you allow negative indices, 
Because I come from Python, and because some (not all) array methods in JS allow them. But it is not important to me. Just thought it would be OK.

> and insert the object at an index different from the one the user requested.
I don't?

> It is too risky.
Would be a good argument, but I don't see why.

> Furthermore, and most importantly, indices change as objects are inserted and removed.
replaceAt does not change any indices. Array insertion (like the proposed addAt) is not yet a default method in Object3D, to my knowledge. Then only _remove_ changes indices. You are right that using remove instead of replacing with a dummy object will shift everything after the removed object. So there will always be a risk that the user loses control over the indices. **but** that is only a problem if the user cares about the indices in the first place. It does not affect users that see _children_ merely as a set.

> Such a method is appropriate as an application-level method where the use-case is restricted and under developer control.
Maybe, but nonetheless it represents the very basic array operation of assignment by index in a conservative and efficient way. Maybe it is not appropriate to think of _children_ as an array, but that is exactly what I do, and it is currently implemented as one.

> Sorry. As I indicated a month ago, I do not believe replaceAtIndex( index, newChild ) is a method we should be implementing.
No big problem if it ends that way. I just hope it is for the right reasons. (Though it doesn't _have to_ be for the right reasons either. ;-))

**Comment by EliasHasle** [comment]
@WestLangley wrote:
> Furthermore, and most importantly, indices change as objects are inserted and removed.

EliasHasle wrote:
> replaceAt does not change any indices. And array insertion (like the proposed addAt) is not yet a default method in Object3D, to my knowledge. Then only remove changes indices. You are right that using remove instead of replacing with a dummy object will shift everything after the removed object.

OK. I get it now. switching the order of _add_s will impact the indices too, and can mess things up if the user cares about indices. This is not a problem unique to this PR (it happens wherever you have lists in programming), but you are right that it adds some risk to the threejs programming, and thus it may not deserve a place in the core.

I will continue using the techniques of this PR in my own applications, whenever I think that is appropriate, but I will close this PR now. Do you think it could/should be resurrected as some kind of plugin? Because I have done some work making it failsafe. If anyone else later wants to apply the same technique, then it may be a benefit that someone has already written the code. :-)
